src/Dotnet.Script.DependencyModel/Context/ScriptDependencyInfoProvider.cs

```diff
@@ -0,0 +1,76 @@
+﻿using System;
+using System.IO;
+using System.Linq;
+using System.Xml.Linq;
+using Dotnet.Script.DependencyModel.Logging;
+using Microsoft.Extensions.DependencyModel;
+
+namespace Dotnet.Script.DependencyModel.Context
+{
+    /// <summary>
+    /// Represents a class that is capable of providing 
+    /// the <see cref="DependencyContext"/> for a given project file (csproj).
+    /// </summary>
+    public class ScriptDependencyInfoProvider 
+    {
+        private readonly IRestorer[] _restorers;
+        private readonly Logger _logger;
+
+        public ScriptDependencyInfoProvider(IRestorer[] restorers, LogFactory logFactory)
+        {
+            _restorers = restorers;
+            _logger = logFactory.CreateLogger<ScriptDependencyInfoProvider>();
+        }
+
+        /// <summary>
+        /// Gets the <see cref="DependencyContext"/> for the project file.
+        /// </summary>
+        /// <param name="pathToProjectFile">The path to the project file.</param>
+        /// <returns>The <see cref="DependencyContext"/> for a given project file (csproj).</returns>
+        public ScriptDependencyInfo GetDependencyInfo(string pathToProjectFile)
+        {            
+            Restore(pathToProjectFile);
+            var context =  ReadDependencyContext(pathToProjectFile);
+            var nugetPackageFolders = GetNuGetPackageFolders(pathToProjectFile);
+            return new ScriptDependencyInfo(context, nugetPackageFolders);
+        }
+
+        private void Restore(string pathToProjectFile)
+        {
+            foreach (var restorer in _restorers)
+            {
+                if (restorer.CanRestore)
+                {
+                    restorer.Restore(pathToProjectFile);
+                    return;
+                }
+            }
+        }
+
+        private DependencyContext ReadDependencyContext(string pathToProjectFile)
+        {
+            _logger.Debug($"Reading dependency context from {pathToProjectFile}");
+
+            var pathToAssetsFiles = Path.Combine(Path.GetDirectoryName(pathToProjectFile), "obj", "project.assets.json");
+
+            using (FileStream fs = new FileStream(pathToAssetsFiles, FileMode.Open, FileAccess.Read))
+            {
+                // https://github.com/dotnet/core-setup/blob/master/src/managed/Microsoft.Extensions.DependencyModel/DependencyContextJsonReader.cs
+                using (var contextReader = new DependencyContextJsonReader())
+                {
+                    return contextReader.Read(fs);
+                }
+            }
+        }
+
+        private string[] GetNuGetPackageFolders(string pathToProjectFile)
+        {
```

Comment:
i think this method may throw in various situations, we should probably try to gracefully exit from here in those cases?

Reply:
Yes, it could throw although something very special must happen in order for it to fail.
I've thought long and hard about what to do with exceptions inside this library. Should we log and eat exceptions or should be just let it bubble up the caller and put the burden of try/catching/log on them? The callers here would be `dotnet.script` and `Omnisharp.Script`. I did not want to do log/rethrow since that would probably cause a duplicate log message, but we could add an exception handler in the `GetDepenencies` method in the dependency resolvers. This handler logs the exception and simply yields an empty list of dependencies. It would save the caller some work.

Reply:
yeah sounds like a good idea. let's leave this for now, I'd like to merge it in, and we do that in the future